jfix-stdlib-ratelimiter/src/test/kotlin/ru/fix/stdlib/ratelimiter/SuspendableRateLimitedDispatcherTest.kt

```diff
@@ -0,0 +1,713 @@
+        }
+
+        trackableDispatcherWithCF.completeAllAndClose()
+    }
+
+    @Test
+    fun `WHEN results arrived THEN indicators are correctly adjusted`() = runBlocking {
+        val profiler = AggregatingProfiler()
+        val reporter = profiler.createReporter()
+        val trackableDispatcher = TrackableDispatcher(profiler)
+
+        trackableDispatcher.submitCompletedTasks(1..4)
+        trackableDispatcher.submitTasks(5..6, 3000) // long executed tasks
+        await().atMost(Duration.ofSeconds(1)).until {
+            trackableDispatcher.isSubmittedTaskInvoked(1..6)
+        }
+        reporter.buildReportAndReset().assertSoftly {
+            it.getMetric("active_async_operations") shouldBe 2
+        }
+
+        trackableDispatcher.submitCompletedTasks(7..10)
+        await().atMost(Duration.ofSeconds(1)).until {
+            trackableDispatcher.isSubmittedTaskInvoked(7..10)
+        }
+        reporter.buildReportAndReset().assertSoftly {
+            it.getMetric("active_async_operations") shouldBe 2
+        }
+
+        await().atMost(Duration.ofSeconds(3)).until {
+            trackableDispatcher.isSubmittedTaskFinished(1..10)
+        }
+        reporter.buildReportAndReset().assertSoftly {
+            it.getMetric("active_async_operations") shouldBe 0
+        }
+
+        trackableDispatcher.completeAllAndClose()
+    }
+
+    /**
+     * When dispatcher closingTimeout is enough for pending tasks to complete
+     * such tasks will complete normally
+     */
+    @Test
+    @Timeout(10, unit = TimeUnit.SECONDS)
+    fun `on shutdown fast tasks with CompletableFuture complete normally`() {
+
+        val dispatch = createDispatcher(closingTimeout = 5_000)
+
+        val futures = List(3) {
+            dispatch.compose {
+                completedFuture(true)
+            }
+        }
+        dispatch.close()
+
+        futures.forEach { future: CompletableFuture<*> ->
+            future.isDone.shouldBeTrue()
+            future.isCompletedExceptionally.shouldBeFalse()
+        }
+    }
+
+    /**
+     * When dispatcher closingTimeout is enough for pending tasks to complete
+     * such tasks will complete normally
+     */
+    @Test
+    @Timeout(10, unit = TimeUnit.SECONDS)
+    fun `on shutdown fast tasks complete normally`() = runBlocking {
+
+        val dispatch = createDispatcher(closingTimeout = 5_000)
+
+        val results = List(3) {
+            dispatch.decorateSuspend {
+                true
+            }
+        }
+        dispatch.close()
+
+        results.forEach { result ->
+            result.shouldBeTrue()
+        }
+    }
+
+    @Test
+    @Timeout(5, unit = TimeUnit.SECONDS)
+    fun `on shutdown slow tasks with CompletableFuture complete exceptionally`() {
+        val dispatch = createDispatcher(closingTimeout = 0)
+
+        for (i in 1..3) {
+            scope.async {
+                dispatch.compose {
+                    sleep(3000)
```

Comment:
do not use blocking calls

```suggestion
                    GlobalScope.future { 
                        delay(3_000)
                        true
                    }
```

Reply:
Fixed